### PR TITLE
feat(create-waku): `--example` option

### PR DESCRIPTION
**Pull request by ojj1123**
related https://github.com/dai-shi/waku/discussions/315

I wrote the prototype for `--example` option. I've referred to [create-next-app](https://github.com/vercel/next.js/tree/canary/packages/create-next-app)
I will be adding `automatically installing dependencies` in next PR.

## What features were added

```
pnpm create waku --example 14_react-tweet // will download https://github.com/dai-shi/waku/tree/main/examples/14_react-tweet

pnpm create waku --example https://github.com/dai-shi/waku/tree/main/examples/14_react-tweet

pnpm create waku --example https://github.com/himself65/waku-vercel-template
```

also in this PR, i've drop --choose-example as @dai-shi said
https://github.com/dai-shi/waku/discussions/315#discussioncomment-8708156

## What I want to discuss
In create-next-app, [they are dealing with the branch name or file path containing the slash.](https://arc.net/l/quote/dsgpummq) But I've thought this case is a bit rare one so I didn't implement this feature in create-waku. How about this case?

### CODE REVIEW

File: packages/create-waku/src/helpers/examples.ts
Comment (dai-shi) [inline packages/create-waku/src/helpers/examples.ts:4]: We use separate import line for types.
```suggestion
import type { ReadableStream } from 'stream/web';
```

File: packages/create-waku/src/index.ts
Comment (dai-shi) [inline packages/create-waku/src/index.ts:18]: The file name "examples" is a bit confusing. Maybe simply like this for now?
```suggestion
} from './helpers';
```

File: packages/create-waku/src/index.ts
Comment (dai-shi) [inline packages/create-waku/src/index.ts:44]: We prefer curly blocks.
```suggestion
  if (exampleToken) {
    return exampleToken.value;
  }
  return;
```

Comment (dai-shi) [reply packages/create-waku/src/index.ts:44]: I prefer this:
```suggestion
  return exampleToken?.value;
```

**Review by dai-shi — COMMENTED** [review]
https://github.com/dai-shi/waku/pull/581/files#diff-995f71dd25297bc098cf21c18a6770843422edd5863adc74f7b181ca7c803bcfR25
We would like to copy only `examples/01_template`.

@himself65 Can you review please?

(Maybe, it's time for us to add some unit tests.)

**Comment by dai-shi** [comment]
> they are dealing with the branch name or file path containing the slash.

Branch name or tag name is actually important because `examples` in `main` branch will be breaking with older versions.

File: packages/create-waku/src/helpers/examples.ts
Comment (himself65) [inline packages/create-waku/src/helpers/examples.ts:4]: use `node:` prefix

File: packages/create-waku/src/helpers/examples.ts
Comment (himself65) [inline packages/create-waku/src/helpers/examples.ts:36]: can you copy some type from github

File: packages/create-waku/src/index.ts
Comment (himself65) [inline packages/create-waku/src/index.ts:95]: Not really necessary in production. But will error if someone use macOS and debug locally

File: packages/create-waku/src/index.ts
Comment (ojj1123) [reply packages/create-waku/src/index.ts:18]: how about this?

```
/helper
  examples.ts
  index.ts

// /helper/index.ts
export * from './examples'
```

or something else?

Comment (dai-shi) [reply packages/create-waku/src/index.ts:18]: that's worse. we don't use `index.ts` except for the root.

Comment (dai-shi) [reply packages/create-waku/src/index.ts:18]:
```
$ ls **/index.ts
examples/02_demo/src/lib/index.ts	packages/website/src/hooks/index.ts
packages/create-waku/src/index.ts	packages/website/src/utils/index.ts
packages/website/src/atoms/index.ts
```
I don't particularly like them either.

We used to use cli.ts in create-waku, but there's a limitation with a bundler. cc @himself65

Comment (ojj1123) [reply packages/create-waku/src/index.ts:18]: I got it.
But I thought it's good to refactor for seperating the util functions into that `helpers` folder because the create-waku's `index.ts` is a bit large
this is why i make the helpers folder.
Do you want to make only `helpers.ts`?

File: packages/create-waku/src/helpers/examples.ts
Comment (ojj1123) [inline packages/create-waku/src/helpers/examples.ts:29]: I take a part of the response type from github docs [here](https://docs.github.com/en/rest/repos/repos?apiVersion=2022-11-28#get-a-repository)

cc. @himself65

**Comment by ojj1123** [comment]
> > they are dealing with the branch name or file path containing the slash.
> 
> Branch name or tag name is actually important because `examples` in `main` branch will be breaking with older versions.

@dai-shi 
what you mean?
you mean `create-waku` support the features below?

```
pnpm create waku --example 01_template // by default, download example from default branch

pnpm create waku --example 01_template --branch with-middleware // download example from with-middleware
```

This PR's implementation is just to download example from `default branch` by default

> because `examples` in `main` branch will be breaking with older versions.

I didn't understand this situation

File: packages/create-waku/src/index.ts
Comment (dai-shi) [reply packages/create-waku/src/index.ts:18]: I'm fine with `helpers/*.ts`. It's just that "examples.ts" is confusing, just a bit. It can be read as "examples of helpers" instead of "helpers for examples."

**Comment by dai-shi** [comment]
> I didn't understand this situation

Let's say, we release v0.20.0. At that point, we can use `npm create --example 01_template`, and it works fine.

Meanwhile, we will start merging PRs for v0.21.0, and examples will only work with the new (unreleased) version. If we run `npm create --example 01_template` at that point, the template won't work with v0.20.0.

File: packages/create-waku/src/index.ts
Comment (ojj1123) [reply packages/create-waku/src/index.ts:18]: i see. i will rename `examples.ts` file to `example-option.ts`

**Comment by ojj1123** [comment]
I got that situation.
but each example have `package.json`. I think that new example for working only with new (unreleased) version is added during or after merging new release.
Wouldn't it be safe if the examples aren't modified before new release?

Still if you think it is nessasary to get the `branch or tag` option, i will make it next PR

**Comment by dai-shi** [comment]
> is added during or after merging new release.

No, that's not the case.
Well, package.json lies in the examples folder (it only works with the monorepo setup.)

> Still if you think it is nessasary to get the branch or tag option, i will make it next PR

Yes, it's necessary, but can be a follow-up PR.

**Comment by ojj1123** [comment]
Please review this PR.
I will do the works next PR below:

- `branch` option for downloading the examples https://github.com/dai-shi/waku/pull/581#issuecomment-1985104060
- Automatically installing dependencies
- writing e2e test code or refactoring

File: packages/create-waku/package.json
Comment (dai-shi) [inline packages/create-waku/package.json:28]: Maybe we can just `cp -r ../../examples/01_template template`. You can consider it in the follow-up PRs.

File: packages/create-waku/src/index.ts
Comment (dai-shi) [inline packages/create-waku/src/index.ts:35]: Maybe this works?
```suggestion
    typeof err.message === 'string'
```

**Review by dai-shi — APPROVED** [review]
I didn't check the behavior though.

File: packages/create-waku/src/index.ts
Comment (Aslemammad) [inline packages/create-waku/src/index.ts:286]: can we use this? https://github.com/antfu/ni

File: packages/create-waku/src/helpers/example-option.ts
Comment (Aslemammad) [inline packages/create-waku/src/helpers/example-option.ts:104]: is this downloading the whole waku repo for users?

File: packages/create-waku/package.json
Comment (ojj1123) [reply packages/create-waku/package.json:28]: yeah i've tried to do that. but i've got it like below:

```
/template
  /src
  package.json
  ...
```

i want to do this below:
```
/template
  /01_template
```

because it will be added some templates 👍

File: packages/create-waku/src/index.ts
Comment (ojj1123) [reply packages/create-waku/src/index.ts:35]: okay i will try that next PR

File: packages/create-waku/src/index.ts
Comment (ojj1123) [reply packages/create-waku/src/index.ts:286]: Oh thanks.
My plan is just below:

```
exec(`${packageManager} install`)
```

i don't know [ni](https://github.com/antfu/ni) but if it is useful, let's use it.

File: packages/create-waku/src/helpers/example-option.ts
Comment (ojj1123) [reply packages/create-waku/src/helpers/example-option.ts:104]: Yes, the only way to get examples of `waku` is to download the whole waku repository.
Do you have any concerns or suggestions?

Comment (Aslemammad) [reply packages/create-waku/src/helpers/example-option.ts:104]: Yea, it might be too much size to download for the user. If there's no other way then let's go with this. but it's better to find a way.

File: packages/create-waku/src/index.ts
Comment (Aslemammad) [reply packages/create-waku/src/index.ts:286]: Yes that can work too. just wanted to drop a possibility.

**Comment by ojj1123** [comment]
@dai-shi 
I have no write access to this repository.
I was wondering if you could merge it 👍

File: packages/create-waku/package.json
Comment (dai-shi) [reply packages/create-waku/package.json:28]: okay, i thought the former is fine (single template). We don't know the future, but for now we have only one template.

File: packages/create-waku/src/index.ts
Comment (dai-shi) [reply packages/create-waku/src/index.ts:286]: Pay special attention when choosing a dependency. we generally avoid adding new dependencies especially if they are new.

**Comment by dai-shi** [comment]
> I was wondering if you could merge it

Basically, it feels like we are still in the middle of review.
We are not confident if there are no bugs, or if it's well designed / coded.
But, I guess you take the responsibility, so let's merge it this time.

> okay i will try that next PR

out of curiosity, why do you prefer separating PRs?

**Comment by ojj1123** [comment]
> Basically, it feels like we are still in the middle of review.
> We are not confident if there are no bugs, or if it's well designed / coded.

yeah i also agree with you. If you have any feedback, please let me know !

> out of curiosity, why do you prefer separating PRs?

i prefer separating the works such as refactoring and other features. I just want to focus on the interests of this PR and i tend to separate the task into smaller tasks and prioritize them.
if you have the other opinion, i could follow your one :)

**Comment by dai-shi** [comment]
> if you have the other opinion, i could follow your one :)

Maybe we can create a branch, and send PRs there. Will let you know when it's appropriate.

Meanwhile, send follow-up PRs to `main`.

**Pull request merged**
